CMake/Autodownload.cmake

```diff
@@ -0,0 +1,84 @@
+function(get_deps LINK DEPS_NAME PACKAGE)
+
+  file(DOWNLOAD ${LINK}
+         "${CMAKE_BINARY_DIR}/deps/${PACKAGE}"
+          STATUS DOWNLOAD_STATUS_LIST LOG DOWNLOAD_LOG
+          SHOW_PROGRESS)
+    list(GET DOWNLOAD_STATUS_LIST 0 DOWNLOAD_STATUS)
+    if (DOWNLOAD_STATUS EQUAL 0)
+      if (NOT ${DEPS_NAME} MATCHES "stb")
+        execute_process(COMMAND ${CMAKE_COMMAND} -E
+            tar xf "${CMAKE_BINARY_DIR}/deps/${PACKAGE}"
+            WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/deps/")
+
+        # Get the name of the directory.
+        file (GLOB DIRECTORIES RELATIVE "${CMAKE_BINARY_DIR}/deps/"
+            "${CMAKE_BINARY_DIR}/deps/${DEPS_NAME}*.*")
+        if(${DEPS_NAME} MATCHES "boost")
+          file (GLOB DIRECTORIES RELATIVE "${CMAKE_BINARY_DIR}/deps/"
+              "${CMAKE_BINARY_DIR}/deps/${DEPS_NAME}*_*")
+        endif()
+        # list(FILTER) is not available on 3.5 or older, but try to keep
+        # configuring without filtering the list anyway (it might work if only
+        # the file ensmallen-latest.tar.gz is present.
+        if (${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.6.0")
+          list(FILTER DIRECTORIES EXCLUDE REGEX ".*\.tar\.gz")
+          list(FILTER DIRECTORIES EXCLUDE REGEX ".*\.tar\.xz")
+        endif ()
+        list(LENGTH DIRECTORIES DIRECTORIES_LEN)
+        message("Print directories: " ${DIRECTORIES})
+        if (DIRECTORIES_LEN GREATER 0)
+          if (${DEPS_NAME} MATCHES "armadillo")
+            list(GET DIRECTORIES 0 ARMADILLO_DIR)
+            set(ARMADILLO_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${ARMADILLO_DIR}/include" CACHE INTERNAL "")
```

Comment:
This code is really specific to each library... it would be better if the standalone function could take in, say, the name of the variable that needs to be set as an argument or something like that.  That way we could remove all code specific to each particular dependency from this function, and have that specific to the place where `get_deps()` is called.  Let me know what you think. :)

Reply:
I know, I have thought about that, but at the end I have to set the `INCLUDE_DIR` for each library. Do you have an idea to do this? is there anyway to return a value from `CMake` functions ?

Reply:
This is the kind of workaround I was thinking could be used here: https://cmake.org/pipermail/cmake/2008-November/025423.html

You just add an additional variable to the function that's a "return variable", then at the calling site, just add another command to set the desired variable name to have the value of the "return variable". :+1:

Reply:
Thanks I was able to find a solution for this one, I did not push anything since I am waiting for the `stb.tar.gz` to get merged